pennylane/gradients/parameter_shift.py

```diff
@@ -0,0 +1,471 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains functions for computing the parameter-shift gradient
+of a qubit-based quantum tape.
+"""
+# pylint: disable=protected-access,too-many-arguments
+import numpy as np
+
+import pennylane as qml
+
+from .finite_difference import finite_diff, generate_shifted_tapes
+
+
+def _get_operation_recipe(tape, t_idx, shift=np.pi / 2):
+    """Utility function to return the parameter-shift rule
+    of the operation corresponding to trainable parameter
+    t_idx on tape.
+
+    If the corresponding operation has grad_recipe=None, then
+    the default two-term parameter-shift rule is assumed.
+    """
+    # get the index of the parameter in the tape
+    parameter_idx = list(tape.trainable_params)[t_idx]
+
+    # get the corresponding operation
+    op = tape._par_info[parameter_idx]["op"]
+
+    # get the corresponding operation parameter index
+    # (that is, index of the parameter within the operation)
+    op_p_idx = tape._par_info[parameter_idx]["p_idx"]
+
+    # return the parameter-shift gradient for that
+    # operation parameter.
+    return op.get_parameter_shift(op_p_idx, shift=shift)
+
+
+def _process_gradient_recipe(gradient_recipe, tol=1e-10):

```

Comment:
It looks like similar logic is used in the `finite_diff_coeffs` so might be something else for a utils module? (but only if it makes sense, and reuse would be possible)

Reply:
Yes, the only difference is that parameter-shift rules can _scale_ the parameter, which finite-differences doesn't.